Importing data File in pandas dataframe

In [663]:
import pandas as pd

ImmiData = pd.read_csv("D:\DQ_WeeklyChallenge1\Data\H-2-certification-decisions.csv")
print(ImmiData.shape)

(213155, 22)


C:\Users\owner\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2723: DtypeWarning: Columns (0,8,9,14,15,16,17,18,20,21) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Printing the first and last rows from the dataframe

In [664]:
print(ImmiData.iloc[0])
last =ImmiData.shape[0] - 1
print(ImmiData.iloc[last])

case_no                                                       9455254
visa_type                                                        H-2B
fy                                                               2000
last_event_date                                            1999-10-01
case_status                                         DENIED, TEMPORARY
n_requested                                                         1
n_certified                                                         0
is_certified                                                    False
certification_begin_date                                          NaN
certification_end_date                                            NaN
job_title                                            PROJECT DIRECTOR
employer_name               CALIFORNIA PUBLIC INTEREST RESEARCH GROUP
employer_state                                                     CA
employer_city                                             LOS ANGELES
employer_address_1  

In [665]:
ImmiData

case_no visa_type    fy last_event_date  \
0                  9455254      H-2B  2000      1999-10-01   
1                  8222219      H-2B  2000      1999-10-04   
2                  8222168      H-2B  2000      1999-10-04   
3                  8222169      H-2B  2000      1999-10-04   
4                  8222170      H-2B  2000      1999-10-04   
5                  8222171      H-2B  2000      1999-10-04   
6                  8222172      H-2B  2000      1999-10-04   
7                  8222173      H-2B  2000      1999-10-04   
8                  8222174      H-2B  2000      1999-10-04   
9                  8222175      H-2B  2000      1999-10-04   
10                 8222177      H-2B  2000      1999-10-04   
11                 8222167      H-2B  2000      1999-10-04   
12                 8222218      H-2B  2000      1999-10-04   
13                 8222179      H-2B  2000      1999-10-04   
14                 8222220      H-2B  2000      1999-10-04   
15                 8222221      H-2B  2000      1999-10-04   
16                 8222222      H-2B  2000      1999-10-04   
17                 8222223      H-2B  2000      1999-10-04   
18                 8222224      H-2B  2000      1999-10-04   
19                 8222225      H-2B  2000      1999-10-04   
20                 8222226      H-2B  2000      1999-10-04   
21                 9455255      H-2B  2000      1999-10-04   
22                 8222217      H-2B  2000      1999-10-04   
23                 8222036      H-2B  2000      1999-10-04   
24                 8222166      H-2B  2000      1999-10-04   
25                 7229433      H-2B  2000      1999-10-04   
26                 7229434      H-2B  2000      1999-10-04   
27                 7229435      H-2B  2000      1999-10-04   
28                 8221912      H-2B  2000      1999-10-04   
29                 8221926      H-2B  2000      1999-10-04   
...                    ...       ...   ...             ...   
213125  H-400-16012-019962      H-2B  2016      2016-03-31   
213126  H-400-16016-774473      H-2B  2016      2016-03-31   
213127  H-400-15363-237825      H-2B  2016      2016-03-31   
213128  H-400-16017-429218      H-2B  2016      2016-03-31   
213129  H-400-16018-716868      H-2B  2016      2016-03-31   
213130  H-400-16015-254415      H-2B  2016      2016-03-31   
213131  H-400-16015-260642      H-2B  2016      2016-03-31   
213132  H-400-16018-278455      H-2B  2016      2016-03-31   
213133  H-400-16015-824805      H-2B  2016      2016-03-31   
213134  H-400-16015-984845      H-2B  2016      2016-03-31   
213135  H-400-16019-056415      H-2B  2016      2016-03-31   
213136  H-400-16019-596644      H-2B  2016      2016-03-31   
213137  H-400-15310-923655      H-2B  2016      2016-03-31   
213138  H-400-16020-185396      H-2B  2016      2016-03-31   
213139  H-400-16020-922154      H-2B  2016      2016-03-31   
213140  H-400-15349-581124      H-2B  2016      2016-03-31   
213141  H-400-15313-531415      H-2B  2016      2016-03-31   
213142  H-400-16007-780014      H-2B  2016      2016-03-31   
213143  H-400-15356-040699      H-2B  2016      2016-03-31   
213144  H-400-16021-735040      H-2B  2016      2016-03-31   
213145  H-400-16021-907386      H-2B  2016      2016-03-31   
213146  H-400-16021-493855      H-2B  2016      2016-03-31   
213147  H-400-16019-848021      H-2B  2016      2016-03-31   
213148  H-400-16022-671731      H-2B  2016      2016-03-31   
213149  H-400-16023-155298      H-2B  2016      2016-03-31   
213150  H-400-16017-706343      H-2B  2016      2016-03-31   
213151  H-400-15356-790738      H-2B  2016      2016-03-31   
213152  H-400-15317-617135      H-2B  2016      2016-03-31   
213153  H-400-16029-215643      H-2B  2016      2016-03-31   
213154  H-400-16086-922027      H-2B  2016      2016-03-31   

                  case_status  n_requested  n_certified is_certified  \
0           DENIED, TEMPORARY            1            0        False   
1        CERTIFIED, TEMPORARY 

2a. visa_type
i. How many different visa types and how many cases for each?

In [666]:

ImmiData["visa_type"].value_counts()
ImmiData["visa_type"].value_counts()

H-2B    122662
H-2A     90493
Name: visa_type, dtype: int64

2b. last_event_date
i. How are date values represented?

In [667]:
ImmiData.info()
type(ImmiData["last_event_date"].iloc[0])

<class 'pandas.core.frame.DataFrame'>
Int64Index: 213155 entries, 0 to 213154
Data columns (total 22 columns):
case_no                     213155 non-null object
visa_type                   213155 non-null object
fy                          213155 non-null int64
last_event_date             205686 non-null object
case_status                 213155 non-null object
n_requested                 141512 non-null float64
n_certified                 213155 non-null int64
is_certified                213155 non-null bool
certification_begin_date    155351 non-null object
certification_end_date      155342 non-null object
job_title                   212671 non-null object
employer_name               213133 non-null object
employer_state              213029 non-null object
employer_city               213076 non-null object
employer_address_1          169205 non-null object
employer_address_2          21596 non-null object
employer_postal_code        168879 non-null object
worksite_state            

str

c. case_status
i. Probably the most important column to understand the different values
since this describes the status of the application.
ii. This column contains many values, which can be simplified further for
analysis. Create a new column with simplified values for easier analysis.

d. n_requested
i. Calculate summary statistics and look for any outliers!

First we find the missing values and their count in the 'n_requested' column

In [668]:
n_requested_null = pd.isnull(ImmiData["n_requested"])
n_requested_null_true = n_requested_null[n_requested_null == True]
n_requested_null_count = len(n_requested_null_true)
n_requested_null_count

71643

We remove the rows with null value in the 'n_requested' column before calculating the mean of that column

In [669]:
n_requested_nonnull_data = ImmiData["n_requested"][n_requested_null == False]
mean = sum(n_requested_nonnull_data) / len(n_requested_nonnull_data)
print(mean)

#Pandas removes the missing values when we use the pandas mean() function
mean_n_requested = ImmiData["n_requested"].mean()
print(mean_n_requested)

#Summary Statistics:
ImmiData["n_requested"].describe()

18.9165159138
18.916515913844762


count    141512.000000
mean         18.916516
std          42.546266
min           0.000000
25%           3.000000
50%           8.000000
75%          19.000000
max        1866.000000
Name: n_requested, dtype: float64

e. N_certified
i. Calculate summary statistics and look for any outliers!

In [670]:
#Pandas removes the missing values when we use the pandas mean() function
mean_n_certified = ImmiData["n_certified"].mean()
print(mean_n_certified)

#Summary Statistics:
ImmiData["n_certified"].describe()

16.053646407543805


count    213155.000000
mean         16.053646
std          39.135363
min           0.000000
25%           2.000000
50%           6.000000
75%          15.000000
max        1866.000000
Name: n_certified, dtype: float64

f. job_title
i. How many different jobs represented?

In [671]:
print(ImmiData["job_title"].value_counts())
len(ImmiData["job_title"].value_counts())

LABORER, LANDSCAPE                                    24595
CLEANER, HOUSEKEEPING                                  6563
FARMWORKERS AND LABORERS, CROP                         5354
FARMWORKER; DIVERSIFIED CROP                           5273
SHEEPHERDER                                            5258
LANDSCAPE LABORER                                      4942
STABLE ATTENDANT                                       4198
FARMWORKER, DIVERSIFIED CROP                           2978
FARMWORKER                                             2965
KITCHEN HELPER                                         2919
SHEEP HERDER                                           2894
CONSTRUCTION WORKER I                                  2651
CONSTRUCTION WORKER II                                 2472
GROUNDSKEEPER, INDUSTRIAL COMMERCIAL                   2457
FARM WORKERS & LABORERS, CROP                          2227
COOK                                                   2213
FARM WORKER; DIVERSIFIED CROPS II       

8752

g. organization_flag
i. What are the different types of organizations and the distribution of
values?

In [672]:
ImmiData["organization_flag"].value_counts()

S                                            25085
Individual Employer                          14143
Association - Filing as Agent (H-2A Only)     3900
H-2A Labor Contractor or Job Contractor       2073
J                                              883
Association - Joint Employer (H-2A Only)       463
A                                              354
M                                              202
Association - Sole Employer (H-2A Only)         19
Name: organization_flag, dtype: int64

3. For fields with many missing values, determine which ones have enough useful values to
keep and which ones you should avoid using in your analysis.

We first calculate the number of missing values and the percentage of missing values in each column. We can assume that the columns with less than 50% of missing values would be useful in our analysis.
So we discard the rest of the columns(with more than 50% missing values) from our analysis.

We write a function null_count to apply for each column from the Immidata dataframe.

In [673]:
def null_count(column):
    # Make a vector that contains True if null, False if not.
    column_null = pd.isnull(column)
    # Create a new vector with only values where the series is null.
    null = column[column_null == True]
    # Return the percentage of null values.
    return ((len(null) * 100) / (ImmiData.shape[0] - 1))


column_null_count = ImmiData.apply(null_count)
print(column_null_count)


case_no                      0.000000
visa_type                    0.000000
fy                           0.000000
last_event_date              3.504039
case_status                  0.000000
n_requested                 33.610910
n_certified                  0.000000
is_certified                 0.000000
certification_begin_date    27.118421
certification_end_date      27.122644
job_title                    0.227066
employer_name                0.010321
employer_state               0.059112
employer_city                0.037062
employer_address_1          20.618895
employer_address_2          89.868827
employer_postal_code        20.771836
worksite_state              29.589405
worksite_city               61.540013
agent_name                  26.844910
organization_flag           77.893448
is_duplicate                64.681404
dtype: float64


Thus based on our assumption above, we don't use the following columns in our analysis:
employer_address_2          89.868827
worksite_city               61.540013
organization_flag           77.893448
is_duplicate                64.681404

Removing the 4 columns from ImmiData yields us the new dataset new_ImmiData below.

In [674]:
new_ImmiData = ImmiData.drop(['employer_address_2','worksite_city','organization_flag','is_duplicate'], axis=1)
new_ImmiData

case_no visa_type    fy last_event_date  \
0                  9455254      H-2B  2000      1999-10-01   
1                  8222219      H-2B  2000      1999-10-04   
2                  8222168      H-2B  2000      1999-10-04   
3                  8222169      H-2B  2000      1999-10-04   
4                  8222170      H-2B  2000      1999-10-04   
5                  8222171      H-2B  2000      1999-10-04   
6                  8222172      H-2B  2000      1999-10-04   
7                  8222173      H-2B  2000      1999-10-04   
8                  8222174      H-2B  2000      1999-10-04   
9                  8222175      H-2B  2000      1999-10-04   
10                 8222177      H-2B  2000      1999-10-04   
11                 8222167      H-2B  2000      1999-10-04   
12                 8222218      H-2B  2000      1999-10-04   
13                 8222179      H-2B  2000      1999-10-04   
14                 8222220      H-2B  2000      1999-10-04   
15                 8222221      H-2B  2000      1999-10-04   
16                 8222222      H-2B  2000      1999-10-04   
17                 8222223      H-2B  2000      1999-10-04   
18                 8222224      H-2B  2000      1999-10-04   
19                 8222225      H-2B  2000      1999-10-04   
20                 8222226      H-2B  2000      1999-10-04   
21                 9455255      H-2B  2000      1999-10-04   
22                 8222217      H-2B  2000      1999-10-04   
23                 8222036      H-2B  2000      1999-10-04   
24                 8222166      H-2B  2000      1999-10-04   
25                 7229433      H-2B  2000      1999-10-04   
26                 7229434      H-2B  2000      1999-10-04   
27                 7229435      H-2B  2000      1999-10-04   
28                 8221912      H-2B  2000      1999-10-04   
29                 8221926      H-2B  2000      1999-10-04   
...                    ...       ...   ...             ...   
213125  H-400-16012-019962      H-2B  2016      2016-03-31   
213126  H-400-16016-774473      H-2B  2016      2016-03-31   
213127  H-400-15363-237825      H-2B  2016      2016-03-31   
213128  H-400-16017-429218      H-2B  2016      2016-03-31   
213129  H-400-16018-716868      H-2B  2016      2016-03-31   
213130  H-400-16015-254415      H-2B  2016      2016-03-31   
213131  H-400-16015-260642      H-2B  2016      2016-03-31   
213132  H-400-16018-278455      H-2B  2016      2016-03-31   
213133  H-400-16015-824805      H-2B  2016      2016-03-31   
213134  H-400-16015-984845      H-2B  2016      2016-03-31   
213135  H-400-16019-056415      H-2B  2016      2016-03-31   
213136  H-400-16019-596644      H-2B  2016      2016-03-31   
213137  H-400-15310-923655      H-2B  2016      2016-03-31   
213138  H-400-16020-185396      H-2B  2016      2016-03-31   
213139  H-400-16020-922154      H-2B  2016      2016-03-31   
213140  H-400-15349-581124      H-2B  2016      2016-03-31   
213141  H-400-15313-531415      H-2B  2016      2016-03-31   
213142  H-400-16007-780014      H-2B  2016      2016-03-31   
213143  H-400-15356-040699      H-2B  2016      2016-03-31   
213144  H-400-16021-735040      H-2B  2016      2016-03-31   
213145  H-400-16021-907386      H-2B  2016      2016-03-31   
213146  H-400-16021-493855      H-2B  2016      2016-03-31   
213147  H-400-16019-848021      H-2B  2016      2016-03-31   
213148  H-400-16022-671731      H-2B  2016      2016-03-31   
213149  H-400-16023-155298      H-2B  2016      2016-03-31   
213150  H-400-16017-706343      H-2B  2016      2016-03-31   
213151  H-400-15356-790738      H-2B  2016      2016-03-31   
213152  H-400-15317-617135      H-2B  2016      2016-03-31   
213153  H-400-16029-215643      H-2B  2016      2016-03-31   
213154  H-400-16086-922027      H-2B  2016      2016-03-31   

                  case_status  n_requested  n_certified is_certified  \
0           DENIED, TEMPORARY            1            0        False   
1        CERTIFIED, TEMPORARY 

4. Use the is_duplicate field to remove duplicates from the dataset.

In [675]:
new_ImmiData["is_duplicate"] = new_ImmiData.duplicated(["case_no"])
print(new_ImmiData["is_duplicate"].sum())
new_ImmiData[new_ImmiData["is_duplicate"] == True]


31718


case_no visa_type    fy last_event_date  \
1492               1288699      H-2B  2000      2000-02-22   
1683               8223563      H-2B  2000      2000-03-06   
2304               5376597      H-2B  2000      2000-03-24   
2506               8223920      H-2B  2000      2000-03-30   
2923               3335171      H-2B  2000      2000-04-14   
3636               4358181      H-2B  2000      2000-05-27   
3818               3338609      H-2B  2000      2000-06-14   
3912              10234565      H-2B  2000      2000-06-26   
5498               4366352      H-2B  2001      2000-11-27   
5736               3350596      H-2B  2001      2000-12-21   
7144               5393692      H-2B  2001      2001-02-09   
7385               3354941      H-2B  2001      2001-02-21   
7694               5394344      H-2B  2001      2001-03-02   
8113               6331324      H-2B  2001      2001-03-16   
8399               9496922      H-2B  2001      2001-03-21   
8449               5394340      H-2B  2001      2001-03-21   
8452               5394341      H-2B  2001      2001-03-21   
8453               5394353      H-2B  2001      2001-03-21   
9042               1307224      H-2B  2001      2001-04-09   
9121               5395358      H-2B  2001      2001-04-11   
9299               5395804      H-2B  2001      2001-04-18   
9312               5395805      H-2B  2001      2001-04-18   
9627               1308794      H-2B  2001      2001-04-30   
9831              10241418      H-2B  2001      2001-05-08   
9922               1309143      H-2B  2001      2001-05-11   
12135              6336985      H-2B  2002      2001-10-26   
12557              6337731      H-2B  2002      2001-11-29   
12665              3367386      H-2B  2002      2001-12-12   
12775              6338607      H-2B  2002      2001-12-18   
12848              4383376      H-2B  2002      2001-12-21   
...                    ...       ...   ...             ...   
209536  H-400-15291-499191      H-2B  2016      2015-12-18   
209537  H-400-15315-288446      H-2B  2016      2015-12-18   
209538  H-400-15254-330266      H-2B  2016      2015-12-18   
209539  H-400-15323-823263      H-2B  2016      2015-12-18   
209540  H-400-15334-770262      H-2B  2016      2015-12-18   
209544  H-400-15328-327769      H-2B  2016      2015-12-21   
209545  H-400-15279-571686      H-2B  2016      2015-12-21   
209546  H-400-15337-539363      H-2B  2016      2015-12-21   
209550  H-400-15327-022481      H-2B  2016      2015-12-22   
209551  H-400-15335-437178      H-2B  2016      2015-12-22   
209552  H-400-15345-031584      H-2B  2016      2015-12-22   
209554  H-400-15356-184973      H-2B  2016      2015-12-23   
209557  H-400-15338-097226      H-2B  2016      2015-12-28   
209558  H-400-15344-615421      H-2B  2016      2015-12-28   
209568  H-400-15300-330881      H-2B  2016      2015-12-29   
209569  H-400-15310-927422      H-2B  2016      2015-12-29   
209570  H-400-15329-817794      H-2B  2016      2015-12-29   
209571  H-400-15336-320070      H-2B  2016      2015-12-29   
209572  H-400-15337-161391      H-2B  2016      2015-12-29   
209573  H-400-15342-785064      H-2B  2016      2015-12-29   
209574  H-400-15342-553523      H-2B  2016      2015-12-29   
209575  H-400-15349-593031      H-2B  2016      2015-12-29   
209576  H-400-15363-715577      H-2B  2016      2015-12-29   
209578  H-400-15323-227743      H-2B  2016      2015-12-30   
209583  H-400-15309-945005      H-2B  2016      2015-12-31   
209584  H-400-15318-666238      H-2B  2016      2015-12-31   
209585  H-400-15320-417840      H-2B  2016      2015-12-31   
209586  H-400-15327-317141      H-2B  2016      2015-12-31   
209784  H-400-15041-859523      H-2B  2016      2016-01-08   
209785  H-400-15050-484451      H-2B  2016      2016-01-08   

                 case_status  n_requested  n_certified is_certified  \
1492    CERTIFIED, TEMPORARY          102          102         True   
1683    CERTIFIED, TEMPORARY    

In [ ]:
5. Since each row doesn’t represent an application for a single worker, any data analyses
done will be on a case level not an individual worker level. Brainstorm ideas to modify
the data itself or your analysis when trying to do data analysis that’s centered on workers
themselves (over the cases / applications).